### PR TITLE
damlc: Add Numeric primitives and type-level nats.

### PR DESCRIPTION
**Pull request by associahedron**
A step towards #2289 in the compiler.

This PR adds support for the nat kind, for type-level natural numbers, and the Numeric primitive type and operations. This PR does not remove or replace the Decimal type, nor does it yet support Numeric literals, nor does it touch the standard libraries, nor hook up DAML programs to the new Numeric types in any way.

### CODE REVIEW

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs [hunk 1]
```diff
@@ -189,15 +190,20 @@ _TOptional = prism' TOptional $ \case
   _ -> Nothing
 
 _TUpdate :: Prism' Type Type
-_TUpdate = prism' TList $ \case
+_TUpdate = prism' TUpdate $ \case
```
Comment (associahedron) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs]: I think this is an unrelated bug fix

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs [hunk 2]
```diff
@@ -189,15 +190,20 @@ _TOptional = prism' TOptional $ \case
...
   TUpdate typ -> Just typ
   _ -> Nothing
 
 _TScenario :: Prism' Type Type
-_TScenario = prism' TList $ \case
+_TScenario = prism' TScenario $ \case
```
Comment (associahedron) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs]: This too.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 3]
```diff
@@ -228,6 +247,8 @@ data BuiltinExpr
   -- Numerical conversion
   | BEInt64ToDecimal           -- :: Int64 -> Decimal, always succeeds since 10^28 > 2^63
   | BEDecimalToInt64           -- :: Decimal -> Int64, only converts the whole part, crashes if it doesn't fit
+  | BEInt64ToNumeric           -- :: ∀(s:nat). Int64 -> Numeric s, crashes if it doesn't fit (TODO: verify?)
```
Comment (remyhaemmerle-da) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: It crashes indeed. This is different from the Decimal version where the result always fit.
you can always double check in the [spec](https://github.com/digital-asset/daml/blob/master/daml-lf/spec/daml-lf-1.rst#conversions-functions)

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs [hunk 4]
```diff
@@ -53,6 +53,10 @@ serializabilityConditionsType world0 _version mbModNameTpls vars = go
       TList typ -> go typ
       TOptional typ -> go typ
       TMap typ -> go typ
+      TNumeric n -> go n
+        -- TODO (#2289): decide if n should be forced to be a static nat
+        -- literal within LF's bounds for the numeric type (0 <= n <= 38).
```
Comment (remyhaemmerle-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: In the first delivered version, we will require the `Numeric` to be applied to a static `nat`. 
This simplified a lot of work on the runtime side. 
For instance, dealing with type variables in java/scala codegen is not trivial at all.  
We may relax this restriction in the future if required.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Type.hs [hunk 5]
```diff
@@ -104,6 +105,7 @@ substitute subst = go (Map.foldl' (\acc t -> acc `Set.union` freeVars t) Set.emp
             in  TForall (v1, k) (go fvSubst1 subst1 t)
         | otherwise -> TForall (v0, k) (go fvSubst0 (Map.delete v0 subst0) t)
       TTuple fs -> TTuple (map (second go0) fs)
+      TNat n -> TNat n
```
Comment (cocreature) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Type.hs]: Think about the extra allocation here with `-O0`! :stuck_out_tongue:

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs (hunk 1 above)
Comment (cocreature) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs]: Can we get tests for these? I’m actually somewhat surprised we didn’t already hit this. It seems to be used in the DAML-LF typechecker and afaict it will error if the prism does not match so it shouldn’t be too hard to make a case that fails without this bugfix.

File: compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs [hunk 6]
```diff
@@ -164,13 +164,11 @@ decodeChoice LF1.TemplateChoice{..} =
...
     <*> mayDecode "templateChoiceUpdate" templateChoiceUpdate decodeExpr
 
--- FixMe: https://github.com/digital-asset/daml/issues/2289
---   drop the `error "Numeric not implemented"` in the following function
 decodeBuiltinFunction :: MonadDecode m => LF1.BuiltinFunction -> m BuiltinExpr
 decodeBuiltinFunction = pure . \case
   LF1.BuiltinFunctionEQUAL_INT64 -> BEEqual BTInt64
   LF1.BuiltinFunctionEQUAL_DECIMAL -> BEEqual BTDecimal
-  LF1.BuiltinFunctionEQUAL_NUMERIC -> error "Numeric not implemented"
+  LF1.BuiltinFunctionEQUAL_NUMERIC -> BEEqualNumeric
```
Comment (cocreature) [inline compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs]: Do we have tests for the decoder anywhere? Maybe some roundtrip tests?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs (hunk 4 above)
Comment (associahedron) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: This seems like a good idea, and lets us check the bound here explicitly.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 7]
```diff
@@ -123,6 +123,20 @@ safetyStep = \case
...
+      BEGreaterEqNumeric  -> Safe 2
+      BEAddNumeric        -> Safe 1
+      BESubNumeric        -> Safe 1
+      BEMulNumeric        -> Safe 1
+      BEDivNumeric        -> Safe 1
+      BEInt64ToNumeric    -> Safe 0
+      BENumericToInt64    -> Safe 0
+      BENumericFromText   -> Safe 1
+      BEToTextNumeric     -> Safe 1
+      BERoundNumeric      -> Safe 1
```
Comment (associahedron) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs]: @hurryabit Just to confirm, type applications aren't counted here, right? Otherwise these numbers would have to be higher. (And so would the numbers for BEFoldl, BEFoldr, etc.) But then `BEMapEmpty` says `Safe 1` but it should be `Safe 0`?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 8]
```diff
@@ -228,6 +247,8 @@ data BuiltinExpr
...
+  | BENumericToInt64           -- :: ∀(s:nat). Numeric s -> Int64, only converts the whole part, crashes if it doesn't fit
```
Comment (hurryabit) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: Do we require the fractional part to be zero?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs (hunk 1 above)
Comment (hurryabit) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Util.hs]: IIRC, we only use them as getters. That's why the problem never showed up. Maybe we can just delete them?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs (hunk 4 above)
Comment (hurryabit) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: I agree. Let's keep it tight at the beginning and loosen it when there's demand.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 7 above)
Comment (hurryabit) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs]: Correct, type applications are not counted. (See the cases for `ETyApp` and `ETyLam` below.) The `Safe 1` for `BEMapEmpty` is wrong and must have slipped through my code review. I'll fix it in a separate PR.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs (hunk 8 above)
Comment (associahedron) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]: The fractional part is truncatedp

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs (hunk 4 above)
Comment (associahedron) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]: Ok! I added a check to make this fixed number between 0 and 38 for now.

File: compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs (hunk 6 above)
Comment (associahedron) [reply compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs]: I couldn't find any. I added the issue #2682 to track this.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Type.hs (hunk 5 above)
Comment (associahedron) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Type.hs]: oh, how awful!
